File: source/Log.cpp
#include "Log.hpp"

namespace GTA
{
	using namespace System;
	using namespace System::Collections::Generic;

	void Log::OnStart()
	{
		LogToFile("", "##########\n\n\n\n##########");
	}

	void Log::Debug(... array<String ^> ^message)
	{
		LogToFile("[DEBUG]", message);
	}
	void Log::Error(... array<String ^> ^message)
	{
		LogToFile("[ERROR]", message);
	}

	void Log::DeleteOldLogs()
	{
		DateTime now = DateTime::Now;
		for each(System::String^ path in IO::Directory::GetFiles(IO::Path::GetDirectoryName(Reflection::Assembly::GetExecutingAssembly()->Location), "*.log"))
		{
			DateTime logDate = DateTime::Parse(path->Substring(path->IndexOf("-") + 1, path->IndexOf(".log") - (path->IndexOf("-") + 1)));
			if ((now - logDate).Days >= maxLogAge){
				IO::File::Delete(path);
			}
		}
	}

	void Log::LogToFile(String ^logLevel, ... array<String ^> ^message)
	{
		DateTime now = DateTime::Now;
		String ^logpath = IO::Path::ChangeExtension(Reflection::Assembly::GetExecutingAssembly()->Location, ".log");

		logpath = logpath->Insert(logpath->IndexOf(".log"), "-" + now.ToString("yyyy-MM-dd"));

		try
		{
			IO::FileStream ^fs = gcnew IO::FileStream(logpath, IO::FileMode::Append, IO::FileAccess::Write, IO::FileShare::Read);
			IO::StreamWriter ^sw = gcnew IO::StreamWriter(fs);

			try
			{
				sw->Write(String::Concat("[", now.ToString("HH:mm:ss"), "] ", logLevel, " "));

				for each (String ^string in message)
				{
					sw->Write(string);
				}

				sw->WriteLine();
			}
			finally
			{
				sw->Close();
				fs->Close();
			}
		}
		catch (...)
		{
			return;
		}
	}
}

File: source/core/Log.cs
//
// Copyright (C) 2015 crosire & kagikn & contributors
// License: https://github.com/scripthookvdotnet/scripthookvdotnet#license
//

using System;
using System.IO;

namespace SHVDN
{
	public static class Log
	{
		public enum Level
		{
			Error,
			Warning,
			Info,
			Debug,
		}

		private static string FilePath => Path.ChangeExtension(typeof(ScriptDomain).Assembly.Location, ".log");

		internal static string FileName => Path.GetFileName(FilePath);

		public static void Clear()
		{
			try
			{
				File.WriteAllText(FilePath, string.Empty);
			}
			catch
			{
				// Ignore exceptions
			}
		}

		public static void Message(Level level, params string[] message)
		{
			WriteToFile(level, message);
			WriteToConsole(level, message);
		}

		internal static void WriteToFile(Level level, params string[] message)
		{
			try
			{
				using (var fs = new FileStream(FilePath, FileMode.Append, FileAccess.Write, FileShare.Read))
				{
					using (var sw = new StreamWriter(fs))
					{
						sw.Write(string.Concat("[", DateTime.Now.ToString("HH:mm:ss"), "] "));

						switch (level)
						{
							case Level.Info:
								sw.Write("[INFO] ");
								break;
							case Level.Error:
								sw.Write("[ERROR] ");
								break;
							case Level.Warning:
								sw.Write("[WARNING] ");
								break;
							case Level.Debug:
								sw.Write("[DEBUG] ");
								break;
						}

						foreach (string str in message)
						{
							sw.Write(str);
						}

						sw.WriteLine();
					}
				}
			}
			catch (Exception ex)
			{
				WriteToConsole(Level.Error, "Failed to write to log file: ", ex.ToString());
			}
		}

		internal static void WriteToConsole(Level level, params string[] message)
		{
			var console = AppDomain.CurrentDomain.GetData("Console") as Console;

			if (console == null)
			{
				return;
			}

			switch (level)
			{
				case Level.Error:
					console.PrintError(string.Join(string.Empty, message));
					break;
				case Level.Warning:
					console.PrintWarning(string.Join(string.Empty, message));
					break;
			}
		}
	}
}
